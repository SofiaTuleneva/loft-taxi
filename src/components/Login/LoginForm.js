import React from "react";
import {useForm, Controller} from "react-hook-form";
import {Button, FormControl, TextField} from "@material-ui/core";
import {fetchLoginRequest} from "../../modules/auth";
import {useDispatch, useSelector} from "react-redux";
import {Redirect, Link} from "react-router-dom";
import {paths} from "../../constants/Paths";

function LoginForm() {
	const auth = useSelector(state => state.auth);
	const dispatch = useDispatch();
	const methods = useForm();
	const {handleSubmit, control, errors} = methods;

	const onSubmit = data => {
		const {email, password} = data;

		dispatch(fetchLoginRequest({
			email,
			password,
		}));
	};

	const errorMessages = {
		email: {
			pattern: <span>Введите email корректно</span>,
			required: <span>Введите email</span>,
		},
		password: {
			required: <span>Введите пароль</span>,
		},
	};

	return auth.isLoggedIn ? <Redirect to={paths.map}/> : (
		<form onSubmit={handleSubmit(onSubmit)} className="login__form">
			<h1 className="form__title">Войти</h1>
			<div className="form__subtitle">
				Новый пользователь? <Link to={paths.signup}>Зарегистрируйтесь</Link>
			</div>
			<div className="input__group">
				<FormControl fullWidth>
					{errors?.email && errorMessages.email[errors.email.type]}
					<Controller
						as={<TextField />}
						control={control}
						id="email"
						type="text"
						name="email"
						placeholder="Email*"
						defaultValue=''
						rules={{
							required: true,
							pattern: /^\S+@\S+$/i,
						}}
					/>
				</FormControl>
			</div>
			<div className="input__group">
				<FormControl fullWidth>
					{errors?.password && errorMessages.password[errors.password.type]}
					<Controller
						as={<TextField />}
						control={control}
						id="password"
						type="password"
						name="password"
						placeholder="Пароль*"
						defaultValue=''
						rules={{
							required: true
						}}
					/>
				</FormControl>
			</div>
			<div className="button__group">
				<Button type="submit" data-testid="submit-button" variant="contained" color="primary">
					Войти
				</Button>
			</div>
			<div className="pending">
				{auth.pending ? ' Загрузка...' : null}
			</div>
			{auth.error ?
				<div data-testid={'error-message'} className="error">
					{auth.error?.error}
				</div>
				: ''}
		</form>
	);
}

export default LoginForm;