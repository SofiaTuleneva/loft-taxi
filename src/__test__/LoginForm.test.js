import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderer from "react-test-renderer";
import LoginForm from '../components/Login/LoginForm';
import configureStore from 'redux-mock-store'
import {Provider} from "react-redux";
import {BrowserRouter} from 'react-router-dom';

jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
    Map: () => ({})
}));

const initialState = {auth: {isLoggedIn: false, error: null}};
const mockStore = configureStore();
let store;
store = mockStore(initialState);

it("Login Form renders correctly", () => {
    const component = renderer.create(
        <Provider store={store}><BrowserRouter><LoginForm /></BrowserRouter></Provider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

// describe('Login Form click submit button', () => {
//     describe('with both empty fields',() => {
//         it('show error message', () => {
//             const {getByTestId} = render(<Provider store={store}><BrowserRouter><LoginForm /></BrowserRouter></Provider>);
//
//             fireEvent.click(getByTestId('submit-button'));
//
//             expect(getByTestId('error-message')).toBeInTheDocument();
//         })
//     });
//
//     describe('with both fulfilled inputs', () => {
//         it('not show error message', () => {
//             const {getByTestId, queryByTestId} = render(<Provider store={store}><BrowserRouter><LoginForm /></BrowserRouter></Provider>);
//             const loginField = getByTestId('login-field');
//             const passwordField = getByTestId('password-field');
//             fireEvent.change(loginField, {target: {value: 'sonjareds@yandex.ru'}});
//             fireEvent.change(passwordField, {target: {value: '123'}});
//
//             fireEvent.click(getByTestId('submit-button'));
//
//             expect(queryByTestId('error-message')).toBeNull();
//         })
//     })
// });

