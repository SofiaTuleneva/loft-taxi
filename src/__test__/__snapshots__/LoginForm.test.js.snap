// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders correctly 1`] = `
<div>
  <form
    className="login__form"
    onSubmit={[Function]}
  >
    <h1
      className="form__title"
    >
      Войти
    </h1>
    <div
      className="form__subtitle"
    >
      Новый пользователь? 
      <a
        className="form__link"
        href="/"
      >
        Зарегистрируйтесь
      </a>
    </div>
    <div
      className="input__group"
    >
      <div
        className="MuiFormControl-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="login"
        >
          Login:
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            data-testid="login-field"
            disabled={false}
            id="login"
            name="login"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Логин"
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="input__group"
    >
      <div
        className="MuiFormControl-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="password"
        >
          Password:
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input MuiInput-input"
            data-testid="password-field"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Пароль"
            required={false}
            type="password"
            value=""
          />
        </div>
      </div>
    </div>
    
    <div
      className="button__group"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        data-testid="submit-button"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Войти
        </span>
      </button>
    </div>
  </form>
</div>
`;
